Update docs to remove references to `Box::leak` and reimplement Debug for bytes interner

lopopolo opened the pull request:
Followup to GH-28.

This decreases the scope of the `bstr` dependency quite a bit which makes it feasible to hide it behind another feature flag.

Thanks for your work on GH-28 @CAD97. I won't block on you taking a look at this, but I would appreciate it if you could leave a comment about whether you feel this followup is sufficient.

lopopolo committed:
Remove Box::leak references from the docs

lopopolo committed:
Add From impls for str::Slice

Remove alloc method, fixup docs

lopopolo committed:
Add From impls for bytes::Slice

Remove alloc method, fixup docs. Slice stores raw [u8] now, no more BStr.

lopopolo committed:
Manually implement Debug for bytes::Slice

By delegating to BStr as was the case when Slice contained BStr references.

CAD97 commented on src/bytes.rs:
```diff
@@ -66,22 +67,40 @@ use std::collections::HashMap;
 ///
 /// # Safety
 ///
 /// Must not be `Clone` or `Copy` because the Drop logic assumes this enum is the
```
`Slice` actually can be `Clone` now, since `Box<_>` conveys unique ownership, and a clone of a `Box` creates a new heap object without touching the existing one.

The safety consideration here is rather that even though `Box` is a "unique owner" of the data, it should not be mutably dereferenced, because `as_static_slice` promises the slice to be valid as long as the `Slice` is not dropped.

<details><summary>Extra pedantic language-lawyering</summary>

(If you want to be _extra_ paranoid, you might require that `Slice` is not moved after `as_static_slice` is called. I think it's highly unlikely that it would be required to not move the `Box` while holding a reference to its contents, but depending on the exact formulation of the borrowing model for Rust that _could_ potentially be a valid interpretation of the status quo. (The `StableDeref` trait relies on this not being true, as does other existing unsafe code in the wild. The benefit of this requirement would be always being able to tag a passed-by-value `Box` pointer with LLVM `@unique`, rather than only once it has been mutably dereferenced.) (The API provided here is still possible to implement soundly even if that restriction is required in the future and not upholdable by the library, it just requires storing `*const _` rather than `Box<_>` to remove the extra information.))

lopopolo replied:
Thanks for the extra pedantic explanation. I think the pedantic version is easy enough to enforce by making an opaque wrapper around the Slice enum and using module visibility tricks.

lopopolo replied:
the result of this train of thought is the `internal` module. `Slice` is generic over the slice type and is wrapped by the `Interned` struct which disallows mutable access to the internals of `Slice`.

CAD97 reviewed (approved):
This looks fine to me. Just one little documentation improvement suggestion.

lopopolo committed:
Prevent mutable access to the owned contents of the Slice enum

See GH-29. Although the Box in Slice::Owned is the unique owner of the slice
contents, we are handling out static references to the interior, which requires
its contents not be moved out.

Enforce this invariant by wrapping the Slice enum in an opaque struct that disallows
mutable access to the interior. Module visibility enforces this as well.

This commit also removes BStr from the bytes interner, replacing its debug impl with
String::from_utf8_lossy. The bstr crate is removed. Intaglio now has zero runtime
dependencies, other than std.

lopopolo committed:
Add docs for which symbol methods generate iterator structs in their docs

Fixup imports.

lopopolo commented:
This PR drops the `bstr` dependency.

lopopolo commented:
This PR makes it possible and easy to build interners for `Path`, `OsStr`, and `CStr`.

lopopolo commented:
Also even for `BStr` if we wanted to.
